sklearn/decomposition/nmf.py

```diff
@@ -80,19 +83,166 @@ def _safe_compute_error(X, W, H):
+        # add full np.sum(np.dot(W, H)) - np.sum(X)
+        res += sum_WH - X_data.sum()
+
+    # Itakura-Saito divergence
+    elif beta == 0:
+        div = X_data / WH_data
+        res = np.sum(div) - np.product(X.shape) - np.sum(np.log(div))
+
+    # beta-divergence, beta not in (0, 1, 2)
+    else:
+        if sp.issparse(X):
+            # slow loop, but memory efficient computation of :
+            # np.sum(np.dot(W, H) ** beta)
+            sum_WH_beta = 0
+            for i in range(X.shape[1]):
+                sum_WH_beta += np.sum(fast_dot(W, H[:, i]) ** beta)
+
+        else:
+            sum_WH_beta = np.sum(WH ** beta)
+
+        sum_X_WH = np.dot(X_data, WH_data ** (beta - 1))
+        res = (X_data ** beta).sum() - beta * sum_X_WH
+        res += sum_WH_beta * (beta - 1)
+        res /= beta * (beta - 1)
+
+    return res
+
+
+def _special_sparse_dot(W, H, X):
+    """Computes np.dot(W, H), only where X is non zero."""
+    if sp.issparse(X):
+        ii, jj = X.nonzero()
+        dot_vals = np.multiply(W[ii, :], H.T[jj, :]).sum(axis=1)
+        WH = sp.coo_matrix((dot_vals, (ii, jj)), shape=X.shape)
+        return WH.tocsr()
+    else:
+        return fast_dot(W, H)
+
+
+def _compute_regularization(alpha, l1_ratio, regularization, n_samples,
+                            n_features):
+    """Compute L1 and L2 regularization coefficients for W and H"""
+    alpha_H = 0.
+    alpha_W = 0.
+    if regularization in ('both', 'components'):
+        alpha_H = float(alpha)
+    if regularization in ('both', 'transformation'):
+        alpha_W = float(alpha)
+
+    l1_reg_W = alpha_W * l1_ratio
+    l1_reg_H = alpha_H * l1_ratio
+    l2_reg_W = alpha_W * (1. - l1_ratio)
+    l2_reg_H = alpha_H * (1. - l1_ratio)
+    return l1_reg_W, l1_reg_H, l2_reg_W, l2_reg_H
+
+
+def _check_string_param(sparseness, solver, regularization, beta_loss):
     allowed_sparseness = (None, 'data', 'components')
     if sparseness not in allowed_sparseness:
         raise ValueError(
             'Invalid sparseness parameter: got %r instead of one of %r' %
             (sparseness, allowed_sparseness))
 
-    allowed_solver = ('pg', 'cd')
+    allowed_solver = ('pg', 'cd', 'mu')
     if solver not in allowed_solver:
         raise ValueError(
             'Invalid solver parameter: got %r instead of one of %r' %
             (solver, allowed_solver))
 
+    allowed_regularization = ('both', 'components', 'transformation', None)
+    if regularization not in allowed_regularization:
+        raise ValueError(
+            'Invalid regularization parameter: got %r instead of one of %r' %
+            (regularization, allowed_regularization))
+
+    # 'mu' is the only solver that handles other beta losses than 'frobenius'
+    if solver != 'mu' and beta_loss not in (2, 'frobenius'):
+        raise ValueError(
+            'Invalid beta_loss parameter: solver %r does not handle beta_loss'
+            ' = %r' % (solver, beta_loss))
+
```

Comment:
Maybe issue a user warning when the mu solver is used in conjunction with the nndsvd init.